sdk/core/core-rest-pipeline/documentation/pipeline.md

```diff
@@ -0,0 +1,220 @@
+# Policies
+
+A pipeline policy manipulates a request as it travels through the pipeline. It is conceptually a middleware that is allowed to modify the request before it is made as well as the response when it is received.
+
+```typescript
+export type SendRequest = (request: PipelineRequest) => Promise<PipelineResponse>;
+
+export interface PipelinePolicy {
+  name: string;
+  sendRequest(request: PipelineRequest, next: SendRequest): Promise<PipelineResponse>;
+}
+```
+
+It is given a string name so that consumers can express dependencies relative to existing policies. The `sendRequest` method uses a middleware pattern, which is familiar to JS developers.
+
+# Pipelines
+
+A `Pipeline` composes several policies and provides the ability to make a request that will flow through the policies in a specific order determined by their requirements. Which allows manipulating each request before and after it is made to the server.
+
+```typescript
+export interface AddPipelineOptions {
+  afterPhase?: PipelinePhase;
+  afterPolicies?: string[];
+  beforePolicies?: string[];
+  phase?: PipelinePhase;
+}
+
+export interface Pipeline {
+  addPolicy(policy: PipelinePolicy, options?: AddPipelineOptions): void;
+  clone(): Pipeline;
+  getOrderedPolicies(): PipelinePolicy[];
+  removePolicy(options: { name?: string; phase?: PipelinePhase }): PipelinePolicy[];
+  sendRequest(httpClient: HttpClient, request: PipelineRequest): Promise<PipelineResponse>;
+}
+```
+
+Dependencies expressed when calling `addPolicy` are not sensitive to order. The following code snippets have identical effects:
```

Comment:
This feels somewhat misleading as we will fall back to input order when there is no other way to order them relative to each other. Perhaps it's sufficient to say that you should never rely on ordering and always express constraints using `AddPipelineOptions` when policies have interactions

Reply:
What about this?

> It is important to never rely on policy insert ordering and always express policy execution constraints using `AddPipelineOptions` when policies have interactions. For example, the following code snippets have identical effects: